text/0000-confidential-token-ids.md

```diff
@@ -0,0 +1,270 @@
+    transaction_3 {token_id: 111, fee: 2},
+    transaction_4 {token_id: 888, fee: 3},
+    transaction_5 {token_id: 888, fee: 4}
+]
+```
+
+The sort order would be:
+
+`transaction_id` | `token_id` | `fee_ratio` multiplier | original fee | `fee_ratio` result
+--- | --- | --- | --- | ---
+transaction_2 | 000 | 6 | 1.5 | 9
+transaction_5 | 888 | 2 | 4 | 8
+transaction_1 | 000 | 6 | 1 | 6
+transaction_3 | 111 | 3 | 2 | 6
+transaction_4 | 888 | 2 | 3 | 6
+
+Note how the two transactions paying more than their minimum fee are sorted to the top, in proportion to how much they are paying above the minimum fee, while all minimum fee transactions have an equal chance of beeing sorted because their `fee_ratio` is equal.
+
+### Performance Impact: Size & Space Analysis
+
+- Each TXO has 6 new bytes allocated in the Amount object for the `masked_token_id` (4 bytes for the value, 2 bytes for Protobuf overhead)
+- The RingCT bulletproof signature increases by at least 96 bytes to contain the Proof of Opening (plus Protobuf overhead).
+- Each block now contains a number of fee outputs scaling linearly with the number of transactions in a block, as opposed to a single fee output aggregated from multiple transactions in a block.
+    - In practice, most clients currently output 2 TXOs, one for the recipient and one for change, unless performing a `split_txo` operation. Output growth on chain is currently dominated by the number of outputs in a transaction, and since blocks typically clear fast enough that there is only one transaction in a block for which to aggregate the fee, this should not have a deleterious effect on the space capacity of the chain, especially given that up to [16 outputs are permitted for every transaction](https://github.com/mobilecoinfoundation/mobilecoin/blob/e74d4c821cb213290975d6d5a9390778661cebfe/transaction/core/src/constants.rs#L17).
+
+### Confidentiality Analysis
+
+The [TxPrefix](https://github.com/mobilecoinfoundation/mobilecoin/blob/a9db0bc7bcac9bac1c1232b194003d02bccaf283/transaction/core/src/tx.rs#L149) contains the `token_id` and the `fee` in the clear. We, and any users of the protocol, need to clearly understand the footprint of this data, and the ways in which an attacker may try to discern the `token_id` of TXOs.
+
+The [TxPrefix](https://github.com/mobilecoinfoundation/mobilecoin/blob/edbc0162cbd647b6605fe817a8753508ca4515a2/transaction/core/src/tx.rs#L149) is included in the [Tx](https://github.com/mobilecoinfoundation/mobilecoin/blob/edbc0162cbd647b6605fe817a8753508ca4515a2/transaction/core/src/tx.rs#L103) that is delivered directly to the enclave through an attested channel at the [client\_tx_propose](https://github.com/mobilecoinfoundation/mobilecoin/blob/2d5190e60f6820cebcd585c19d16cecf9ba4d89b/consensus/enclave/impl/src/lib.rs#L299) endpoint. By attested, this means that the client would not deliver this payload unless the enclave's signed measurement matched the client's expectations.
```

Comment:
```suggestion
The [TxPrefix](https://github.com/mobilecoinfoundation/mobilecoin/blob/edbc0162cbd647b6605fe817a8753508ca4515a2/transaction/core/src/tx.rs#L149) is included in the [Tx](https://github.com/mobilecoinfoundation/mobilecoin/blob/edbc0162cbd647b6605fe817a8753508ca4515a2/transaction/core/src/tx.rs#L103) that is delivered directly to the enclave through an attested channel at the [client_tx_propose](https://github.com/mobilecoinfoundation/mobilecoin/blob/2d5190e60f6820cebcd585c19d16cecf9ba4d89b/consensus/enclave/impl/src/lib.rs#L299) endpoint. By attested, this means that the client would not deliver this payload unless the enclave's signed measurement matched the client's expectations.
```

Reply:
The backslash was necessary for rendering in my markdown editor - I'll verify in Github's rendering